classes/actions.php

```diff
@@ -147,8 +147,16 @@ public static function duplicate(array $modules, $sectionnumber = false): void {
         // sorted by their id:
         // Let order of mods in a section be mod1, mod2, mod3, mod4, mod5. If we duplicate mod2, mod4, the order afterwards will be
         // mod1, mod2, mod3, mod4, mod5, mod2(dup), mod4(dup).
+        $cms = [];
+        $errors = [];
         foreach ($idsincourseorder as $cmid) {
-            $duplicatedmod = duplicate_module($modinfo->get_course(), $modinfo->get_cm($cmid));
+            try {
+                $duplicatedmod = duplicate_module($modinfo->get_course(), $modinfo->get_cm($cmid));
+            } catch (\Exception $e) {
+                $errors[$cmid] = 'cmid:' . $cmid . '(' . $e->getMessage() . ')';
```

Comment:
any particular reason why you don't fir an event here, but instead later looping through the errors?

Reply:
I moved the event to catch block.